File: DataStructures/LinkedLists/LinkedList.rb
=begin
* A straightfoward implementation of a classic Singly Linked List data
* structure in Ruby.
*
* www.softwareontheshore.com
=end

class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class LinkedList
  attr_accessor :length, :head

  def initialize()
    @head = nil
    @length = 0
  end

  def add(value)
    current = @head
    new_node = Node.new(value)

    if current === nil
      @head = new_node

      return new_node
    end

    while current.next != nil
      current = current.next
    end

    current.next = new_node
    @length += 1

    return new_node
  end

  def position(value)
    current = @head
    count = 0

    while current.next != nil
      if current.data === value
        return count
      end

      current = current.next
      count += 1
    end

    if current.data = value
      return count
    end

    return -1
  end

  def find(position)
    current = @head
    length = @length
    count = 0

    if position >= length || position < 0
      return -1
    end

    while count < position
      corrent = current.next
      count += 1
    end

    return current
  end

  def remove(position)
    current = @head
    length = @length
    deleted_node = nil
    previous_node = nil
    count = 0

    if position < 0 || position > length
      return -1
    end

    if position === 0
      @head = current.next
      deleted_node = current
      @length -= 1

      return deleted_node
    end

    while count < position
      previous_node = current
      current = current.next
      count += 1
    end

    if previous_node != nil
      previous_node.next = current.next
    end

    deleted_node = current
    current = nil
    @length -= 1

    return deleted_node
  end
end

File: DataStructures/LinkedLists/LinkedList.py
"""
 * A straightfoward implementation of a classic Singly Linked List data
 * structure in Python.
 *
 * www.softwareontheshore.com
"""

class Node(object):
    def __init__(self, data = None, next = None):
        self.data = data
        self.next = next

class LinkedList(object):
    def __init__(self, head = None):
        self.head = head
        self._length = 0

    def __str__(self):
        string = ''
        node = self.head

        if node is not None:
            while node.next is not None:
                string += str(node.data) + ', '
                node = node.next

            string += str(node.data)

        return string

    ## append a new node
    def add(self, data):
        new_node = Node(data)
        current_node = self.head

        if current_node is None:
            self.head = new_node
            self._length += 1
            return new_node

        while current_node.next:
            current_node = current_node.next

        current_node.next = new_node
        self._length += 1

        return new_node

    ## find position of value
    def position(self, value):
        current_node = self.head
        position = 0

        if current_node is not None:
            while current_node.next is not None:
                if current_node.data == value:
                    return position

                current_node = current_node.next
                position += 1

            if current_node.data == value:
                return position

        return -1

    ## find value at position
    def find(self, position):
        current_node = self.head
        length = self._length
        count = 0

        if position >= length or position < 0: return -1

        if current_node is not None:
            while count < position:
                current_node = current_node.next
                count += 1

            return current_node.data

        return -1

    ## remove the node at position
    def remove(self, position):
        current_node = self.head
        length = self._length
        count = 0
        previous_node = None
        deleted_node = None

        if position >= length or position < 0: return -1

        if position == 0:
            self.head = current_node.next
            deleted_node = current_node
            self._length -= 1

            return deleted_node

        if current_node is not None:
            while count < position:
                previous_node = current_node
                current_node = current_node.next
                count += 1

            previous_node.next = current_node.next
            deleted_node = current_node
            current_node = None
            self._length -= 1

            return deleted_node
